### PR TITLE
Desktop: Added arrows to go front and back through note history

### PR DESCRIPTION
**Pull request by naviji**
Add forward and backward arrows to go through note history.

ezgif com-video-to-gif

#2409

### CODE REVIEW

**Comment by laurent22** [comment]
Note: we don't review incomplete PRs.

**Comment by naviji** [comment]
Oops. Sorry. I'll remember that :)

**Comment by naviji** [comment]
Done! Take a look @laurent22 
I want to add some tests too. Any suggestions?

**Comment by nr458h** [comment]
Hi @naviji and welcome to the Joplin-Community. I saw your PR in the forum and was interested.

So I just want to bring up my opinion about the position of the arrows. (I'm no designer nor developer) But I think the arrows shouldn't be placed there for some reason
- In the viewer mode there is no need of the toolbar (there is also an issue for that) and without toolbar no going forwards and backwards in the viewer mode
- this is a toolbar so I would expect clicking this button that some markdown magic is inserted

A suggestion could be - in my opinion - next to the search bar as this functions are somewhat related (I search for a note and want to go back to the previous)

Thanks anyway and the last word belongs to @laurent22  and @PackElend

**Comment by Rishgod** [comment]
@naviji I think you should squash your commits into fewer commits

**Comment by naviji** [comment]
@Rishgod  Yes, I'll do that. But I want to wait until I get a review first.

**Comment by naviji** [comment]
@laurent22 I've implemented all the functionality requested in the issue. Please take a look.

**Comment by laurent22** [comment]
It's a bit hard to follow the PR and I don't understand what "preHistoryNotes" means. History is always "previous" by definition. As you've noticed there's also already an "historyNotes" state so I'm concerned we might be duplicating state here. Could you explain your reasoning please?

**Comment by naviji** [comment]
Sorry. I should have named it better. It was supposed to be called **prevHistoryNotes**.  Still not a good name, I know.

https://stackoverflow.com/questions/6869476/how-to-implement-back-and-forward-functionality-like-browser
historyNotes is the back stack and prevHistoryNotes is the next stack in this recipe for implementing the backward and forward jumps. 

Should I rename it like this?
historyNotes -> previousNotes
preHistoryNotes -> nextNotes

File: ReactNativeClient/lib/reducer.js
Comment (naviji) [inline ReactNativeClient/lib/reducer.js]: This is the part that maintains the two stacks when navigating between notes.

**Comment by laurent22** [comment]
Thanks for the explanation @naviji, it makes sense. Based on the link you've provided and explanation, please could you rename it like below?

historyNotes => backwardHistoryNotes
prevHistoryNotes => forwardHistoryNotes

File: ElectronClient/gui/MainScreen.jsx
Comment (naviji) [inline ElectronClient/gui/MainScreen.jsx]: historyNoteAction contains the **id** and **parent_id** of the note we are navigating away from. 
Maybe it should also be renamed? Along with the function getNoteAction. Suggestions?

**Comment by naviji** [comment]
I'd also like to add some tests, though I'm not sure where to start.

**Comment by naviji** [comment]
I did the renaming. I also did a rebase and squashed everything.
Should I add some tests or additional comments?

**Comment by naviji** [comment]
I've refactored the code. It should be more readable now.

**Comment by mic704b** [comment]
> Should I add some tests

Yes, before this feature is merged, there should be full tests added.

For example, 

For the reducer `handleItemDelete` and `changeSelected` folder, there should be unit tests added to the `reducer.js` unit test file.

For `HistoryHelper` there should be unit tests added to new `services_HistoryHelper.js` file.  (Actually is this really a service? it may be this needs to be somewhere else)

For the changes to reducer function itself, and overall feature testing, there should be new integration tests created in `integration_ForwardBackwardNoteHistory.js`.

**Review by laurent22 — CHANGES_REQUESTED** [review]
Thanks for the update @naviji. The code is well written and I didn't notice any big issue. Could you also please add the tests as asked by @mic704b?

File: ReactNativeClient/lib/reducer.js
Comment (naviji) [reply ReactNativeClient/lib/reducer.js]: It should just be `return newState`

`JSON.stringify(newState.selectedNoteIds) === JSON.stringify(noteIds)` will always be true since we are assigning `newState.selectedNoteIds = noteIds` before the if condition.

I removed `if (JSON.stringify(newState.selectedNoteIds) === JSON.stringify(noteIds)) return state;` because of this situation:

Suppose we are in note1. We then search for note1 again. Now when we press the back button, 
`JSON.stringify(newState.selectedNoteIds) === JSON.stringify(noteIds)` will be true and the old state will be returned without doing the pop operation. 

The history operations should be performed irrespective of `JSON.stringify(newState.selectedNoteIds) === JSON.stringify(noteIds)`

**Comment by naviji** [comment]
@laurent22 I added the unit tests. I'm kind of stuck on the integration tests. It gives me weird random errors. I made a post about it in the forum.

https://discourse.joplinapp.org/t/integration-showallnotes-test-failing-randomly/6664

**Comment by laurent22** [comment]
Do the errors show up if you remove the tests you've added?

**Comment by naviji** [comment]
Yes. I did the testing on a new cloned copy from the original master branch. It still randomly failed when I just ran the integration tests.

**Comment by naviji** [comment]
@laurent22 Finished adding integration tests.

File: ElectronClient/.gitignore
Comment (laurent22) [inline ElectronClient/.gitignore]: Please remove this file from the commit. There's no longer an "app" directory (and if you have one in your workspace, you should remove it).

**Comment by laurent22** [comment]
Thanks @naviji, I've left a few comments on the code.

Another small issue is that it appears the "historyAction" most of the time is "goto". There's only one case where it's "pop" and another where it's "push". In that case, could you remove all the instances where you set it to "goto", and just make the default "goto"?

**Comment by naviji** [comment]
> could you remove all the instances where you set it to "goto," and just make the default "goto"?

I can't do that because I need to distinguish between user-initiated NOTE_SELECT and others. For example, If I'm in note1 and I search for note2, when I use the back arrow, multiple NOTE_SELECT actions are dispatched. These are not initiated by the user directly and shouldn't cause changes in history, like making forwardHistoryNotes empty.

**Comment by PackElend** [comment]
> There's only one case where it's "pop"

btw how can I understand pop, goto is clear, push can makes sense, how to explain pop?

**Comment by laurent22** [comment]
Ok all good now, thanks @naviji!

**Comment by naviji** [comment]
> > There's only one case where it's "pop"
> 
> btw how can I understand pop, goto is clear, push can makes sense, how to explain pop?

Look at this [post](https://stackoverflow.com/questions/6869476/how-to-implement-back-and-forward-functionality-like-browser). 
_push_ and _pop_ refer to what action needs to take place on the backStack.

**Comment by tessus** [comment]
@naviji I haven't reviewed the code, but I was wondering what the limit was.

e.g. I have Joplin running for weeks before I install a new version and restart. What would happen, if I switched between 100,000 notes in that time? Just an exaggeration, but you'll see where I'm going with this. Will it really add 100,000 entries onto a stack? Also, is there some sort of optimization? e.g. if I switched between 2 notes for a few (N) times, would only be those 2 notes on the stack, or the notes * N switches?

File: ReactNativeClient/lib/reducer.js
Comment (mic704b) [inline ReactNativeClient/lib/reducer.js]: More generally, it gets the current note

File: ReactNativeClient/lib/reducer.js
Comment (mic704b) [inline ReactNativeClient/lib/reducer.js]: Missing explicit return for when no note found

File: ReactNativeClient/lib/reducer.js
Comment (mic704b) [inline ReactNativeClient/lib/reducer.js]: Missing test

**Comment by tessus** [comment]
@mic704b the code has already been merged

**Comment by mic704b** [comment]
> @mic704b the code has already been merged

My mistake, I missed that.

**Comment by mic704b** [comment]
BTW there is also an issue if an un-selected note is deleted.

**Comment by mic704b** [comment]
> @mic704b the code has already been merged

> My mistake, I missed that.

It's strange it lets me enter new review comments after merge

File: ReactNativeClient/lib/reducer.js
Comment (laurent22) [reply ReactNativeClient/lib/reducer.js]: Good catch, and that makes me think I'll add an eslint rule for explicit returns as I believe there's one.

**Comment by laurent22** [comment]
Those are good points regardless. @naviji, please could you check Mic's comments and see if you can address them?

**Comment by laurent22** [comment]
> e.g. I have Joplin running for weeks before I install a new version and restart. What would happen, if I switched between 100,000 notes in that time?

I guess we could limit the history to 100 notes or so, what you think @naviji?

> e.g. if I switched between 2 notes for a few (N) times, would only be those 2 notes on the stack, or the notes * N switches?

When I've tested I noticed that it would keep pushing the notes on the stack. That could be improved, for example by having a function that detects sequences in the history and remove successive duplicates, but it could be for another PR

**Comment by tessus** [comment]
> I guess we could limit the history to 100 notes or so, what you think @naviji?

I think 1,000 would be fine too. Not that someone would really go back 1,000 notes. I only meant that there should be a limit. I actually don't really care what it is. For me 20 is probably enough, others might need more. 50, 100, 250. I'm fine with any number.

> When I've tested I noticed that it would keep pushing the notes on the stack. That could be improved, for example by having a function that detects sequences in the history and remove successive duplicates, but it could be for another PR

A new PR is perfectly fine. It also allows for another `good first issue`. I just wanted to bring this up, that's all.

**Comment by tessus** [comment]
@mic704b 

> It's strange it lets me enter new review comments after merge

I think anything is fair game as long as the issue hasn't been locked. But even then collaborators can add/edit comments.

File: ReactNativeClient/lib/reducer.js
Comment (mic704b) [reply ReactNativeClient/lib/reducer.js]: Sounds good. The `eqeqeq` rule might be worth a look too.

Comment (laurent22) [reply ReactNativeClient/lib/reducer.js]: Yes, I didn't include it because when I started I didn't always use `===` and I'm concerned that changing everything to it now might create subtle bugs. New code should indeed use strict equality/inequality. Might still be worth reviewing at some point if we can enable it though.

File: ReactNativeClient/lib/reducer.js
Comment (naviji) [reply ReactNativeClient/lib/reducer.js]: I'll add that. Also a test for Goto Anything.

**Comment by tessus** [comment]
@naviji whatever you do, do not use this branch. Please work off a new branch from the latest git master

**Comment by naviji** [comment]
>BTW there is also an issue if an un-selected note is deleted.

@mic704b  How can I replicate it?

> we could limit the history to 100 notes or so,

@laurent22 Firefox has a [default value](https://support.mozilla.org/en-US/questions/1054076) of 50 for its back button. I'd say 250 is more than enough.

>if I switched between 2 notes for a few (N) times, would only be those 2 notes on the stack, or the notes * N switches?

> having a function that detects sequences in the history and remove successive duplicates

@tessus Firefox and google chrome do no such optimization when clicking between two links repeatedly. I think it wouldn't be worth the trouble.

File: ReactNativeClient/lib/reducer.js
Comment (naviji) [reply ReactNativeClient/lib/reducer.js]: So no explicit return statement is bad style. Got it.

**Comment by tessus** [comment]
@naviji 

>  I'd say 250 is more than enough.

Maybe we should use less then?

> Firefox and google chrome do no such optimization when clicking between two links repeatedly. I think it wouldn't be worth the trouble.

Well, we have a slight problem though. Our situation is different than on browsers. As soon as a we switch to a different notebook, a note is always selected automatically. Therefore the list we build has potentially notes in there, which we actually do not want. But this is probably not something we can fix.
However, I think this PR caused another issue: #2803

**Comment by mic704b** [comment]
> How can I replicate it?

1. Make a folder with some notes
2. Navigate through them to make some history
3. Select a note
4. Right-click another note and delete it (but do not select it)
5. Use the back button to step backwards through the history

Observed: 
When you navigate back over where the note was, the focus disappears completely, as if we have navigated to a hidden note.

Expected:
The navigation should skip the missing note and jump to the next on in the history.

**Comment by naviji** [comment]
@mic704b 
I can't replicate it.
5e747d5269eff010901835

**Comment by mic704b** [comment]
> I can't replicate it.

Neither can I anymore!  I should have noted the commit number.  Sorry.

I do notice a different issue though, the history doesn't seem to track keyboard navigation, ie when I move around the note list using the keyboard.

**Comment by steenstra** [comment]
Is it possible to link these buttons to the back and forward buttons of my mouse?

**Pull request merged**
